### PR TITLE
feat: op consumer chain slashing

### PR DESCRIPTION
**Pull request by parketh**
## Summary

This PR implements slashing for op stack consumer chain.

It updates the cw contract as follows:

- allow FPs to equivocate by submitting mutiple signatures for a block
- add new state to index voted block hashes and equivocation evidences
- add new handlers to slash finality providers and send notifying message to Babylon,  similar to the Cosmos consumer chain implementation

It also updates and add a new test case. We need to add additional integration tests to check the e2e flow with Babylon node.

## Test plan
```bash
# build the contract
cargo build
cargo optimize

# test it
cargo test
cargo integration
```

### CODE REVIEW

File: contracts/op-finality-gadget/src/ibc.rs
Comment (SebastianElvis) [inline contracts/op-finality-gadget/src/ibc.rs:1]: Hmm why would this smart contract support IBC?

Babylon contract needs IBC because it is deployed on consumer chains and needs to establish IBC channels with Babylon. The finality gadget contract does not have this requirement

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (SebastianElvis) [inline contracts/op-finality-gadget/src/exec/finality.rs:375]: The slashing message is not sent via an IBC channel; rather it's by using a message in `res`. This message is sent from the contract to the Cosmos SDK level of Babylon. Babylon side needs a handler for handling this message. 

So in order to complete the slashing user story, we will need to develop the following

- In Babylon node, adding a new message handler that upon the evidence, slash the given FP
- In the contract, getting the evidence of an equivocating FP, and sending it to the above message handler

@maurolacy Do you have any example of CosmWasm-Cosmos SDK communication in hand?

File: contracts/op-finality-gadget/src/ibc.rs
Comment (bap2pecs) [reply contracts/op-finality-gadget/src/ibc.rs:1]: yeah sorry I didn't provide the context. our op contract should be much simpler since we don't need to have most of those logic

File: contracts/op-finality-gadget/src/state/config.rs
Comment (lesterli) [inline contracts/op-finality-gadget/src/state/config.rs:16]: these two fields are unnecessary, we can query them with `consumer_id` from Babylon if need

File: contracts/op-finality-gadget/src/state/config.rs
Comment (lesterli) [inline contracts/op-finality-gadget/src/state/config.rs:17]: what is `babylon` field? it is better to add a comment

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (parketh) [reply contracts/op-finality-gadget/src/exec/finality.rs:375]: ah got it, so because the contract is hosted on Babylon, we cant directly message the node rather than use IBC which is more for cross-chain comms with Cosmos consumers. some code examples would def be helpful 🙏

for the new message handler on Babylon, do you guys have any preference on the name and route of this new handler fn? I can start a PR to add it. 

also, do you think we need to amend the `BTCSlasher` program, or we can simply emit the equivocation evidence inside an `MsgAddFinalitySig` as for babylon fps?

Comment (SebastianElvis) [reply contracts/op-finality-gadget/src/exec/finality.rs:375]: > some code examples would def be helpful 

Hmm actually you can search for `add_message` or `add_submessage` in the code base. Both functions are used for sending a message back to the chain that hosts the smart contract (in this context Babylon). One concrete example is [here](https://github.com/babylonlabs-io/babylon-contract/blob/5f5ba8de0f04fc807a2c5683ab49235c8514cfc9/contracts/babylon/src/ibc.rs#L178-L197)

> for the new message handler on Babylon, do you guys have any preference on the name and route of this new handler fn? I can start a PR to add it.

Thanks, wdyt we create a message for handling equivocation evidences under `x/btcstaking`? Say something like 

```protobuf
rpc EquivocationEvidence(MsgEquivocationEvidence) returns (MsgEquivocationEvidenceResponse);
```

And then upon this message, we verify the evidence, and then call [SlashConsumerFinalityProvider](https://github.com/babylonlabs-io/babylon/blob/4898d8ddd62cb01b4908ef9291a85a11d6e37149/x/btcstaking/keeper/finality_providers.go#L122-L123) to slash this FP.

> also, do you think we need to amend the BTCSlasher program, or we can simply emit the equivocation evidence inside an MsgAddFinalitySig as for babylon fps?

This is probably not needed, although we might want to move the logic [here](https://github.com/babylonlabs-io/babylon/blob/4898d8ddd62cb01b4908ef9291a85a11d6e37149/x/finality/keeper/msg_server.go#L310-L319) to `SlashConsumerFinalityProvider`, such that `SlashConsumerFinalityProvider` will propagate slashing events to relevant consumers, and emit slashing events for the BTC staking tracker.

Comment (bap2pecs) [reply contracts/op-finality-gadget/src/exec/finality.rs:375]: > This is probably not needed

why is that? I thought we will these two steps:

- emit `EventSlashedFinalityProvider` in the new msg handler `EquivocationEvidence`
- modify BTCSlasher to subscribe to this msg and filter the event

Comment (SebastianElvis) [reply contracts/op-finality-gadget/src/exec/finality.rs:375]: If the event is still in type `EventSlashedFinalityProvider`, then BTC slasher will recognise it just like slashed Babylon FP, right?

Comment (bap2pecs) [reply contracts/op-finality-gadget/src/exec/finality.rs:375]: > If the event is still in type `EventSlashedFinalityProvider`, then BTC slasher will recognise it just like slashed Babylon FP, right?

from our discussion in https://www.notion.so/snapchain/OP-Stack-CC-Slashing-Spec-145a62189ff480cbb2bace2403e3c33c:

"The BTCSlasher program subscribes to `EventSlashedFinalityProviderevent` inside the [MsgAddFinalitySig](https://github.com/babylonlabs-io/vigilante/blob/main/btcstaking-tracker/btcslasher/slasher.go#L25) msgs from Babylon node." But this msg is only for Babylon.

but my understanding is we will add a new msg for `EquivocationEvidence`. so it won't be detected by BTCSlasher until we modify it.

am I missing anything?

Comment (bap2pecs) [reply contracts/op-finality-gadget/src/exec/finality.rs:375]: @SebastianElvis we also see there is a `x/btcstkconsumer` module. is it a better place for slashing consumer FPs or do you think `x/btcstaking` is still a better choice?

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (SebastianElvis) [inline contracts/op-finality-gadget/src/exec/finality.rs:326]: Extracting BTC SK does not seem necessary here

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (SebastianElvis) [inline contracts/op-finality-gadget/src/exec/finality.rs:339]: Do we need to emit an event given that Babylon will do this instead?

**Review by SebastianElvis — APPROVED** [review]
Lgtm after some minor comments and merging the PR on Babylon side :+1:

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (parketh) [reply contracts/op-finality-gadget/src/exec/finality.rs:339]: i think its probably still useful to emit an event here. could be for debugging or building a slashing monitor in future.

**Comment by parketh** [comment]
merging this after resolving comments, we can reopen with new PR if there are any other issues (e.g. after implementing some e2e tests)

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (SebastianElvis) [inline contracts/op-finality-gadget/src/exec/finality.rs:311]: WE do not extract SK here

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (SebastianElvis) [inline contracts/op-finality-gadget/src/exec/finality.rs:366]: Probably the syntax needs to be fixed as per https://github.com/babylonlabs-io/babylon/pull/327

**Review by SebastianElvis — COMMENTED** [review]
Lgtm apart from some minor stuff

File: contracts/op-finality-gadget/src/exec/finality.rs
Comment (parketh) [reply contracts/op-finality-gadget/src/exec/finality.rs:311]: sorry what do you mean by this?

**Pull request merged**
